pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -3605,19 +3605,54 @@ public String clientSourceAddressAndPort() {
                     return connectionCheckInProgress;
                 } else {
                     final CompletableFuture<Optional<Boolean>> finalConnectionCheckInProgress =
                             new CompletableFuture<>();
                     connectionCheckInProgress = finalConnectionCheckInProgress;
                     ctx.executor().schedule(() -> {
-                        if (finalConnectionCheckInProgress == connectionCheckInProgress
-                                && !finalConnectionCheckInProgress.isDone()) {
+                        if (!isActive()) {
+                            finalConnectionCheckInProgress.complete(Optional.of(false));
+                            return;
+                        }
+                        if (finalConnectionCheckInProgress.isDone()) {
+                            return;
+                        }
+                        if (finalConnectionCheckInProgress == connectionCheckInProgress) {
+                            /**
+                             * {@link #connectionCheckInProgress} will be completed when
+                             * {@link #channelInactive(ChannelHandlerContext)} event occurs, so skip set it here.
+                             */
                             log.warn("[{}] Connection check timed out. Closing connection.", this.toString());
                             ctx.close();
+                        } else {
+                            /**
+                             * Scenarios that changing {@link #connectionCheckInProgress}.
+                             *   1. {@link #connectionCheckInProgress} will be changed to "null" after a successful
+                             *     "Ping & Pong"
+                             *   2. {@link #connectionCheckInProgress} will be set to a new future the next time
+                             *     {@link #checkConnectionLiveness()} when it is null.
+                             *
+                             * Once {@link #connectionCheckInProgress} is not equal to
+                             *   {@link #finalConnectionCheckInProgress}, it means Scenario 1 occurred before. If
+                             *   "receiving Pong" and "the current scheduled task" are executing at the same time,
+                             *   this log will be printed. Since the two events will be executing at the same thread,
+                             *   the concurrency scenario can not occur, so this log's level is "ERROR".
+                             */
+                            log.error("[{}] Connection check might be success, because the variable"
+                                    + " connectionCheckInProgress has been override by the following check. But this"
+                                    + " scenario is not expected",
+                                    this.toString());
+                            finalConnectionCheckInProgress.complete(Optional.of(true));
```

Comment:
> Do you see a case where that wouldn't be satisfied?

I never find something went wrong related to this case, just add a log 😂 , it should never be printed.

Reply:
I think that this is overly defensive code when there's no assumptions that this code would ever get executed. Such code and comments just add unnecessary clutter to the code base.

Reply:
> I think that this is overly defensive code.

Agree with you, how about just leaving the code there? Remove it after a period of stable operation, because more logs are more helpful for troubleshooting.

Reply:
> Agree with you, how about just leaving the code there? Remove it after a period of stable operation, because more logs are more helpful for troubleshooting.

I think that this code block is just cluttering the code It would be fine if there would be a simple log message without the current overly verbose code comments which could become obsolete in the future if code gets changed. It could be simply something brief like `log.error("[{}] Reached unexpected code block. Completing connection check.", this.toString())` etc.

Reply:
Sure, improved the log and removed the code comments, please take a look again.